File: runtime-go/state/checkpoint.go
// Package state / checkpoint.go —— Checkpoint 与恢复(ch09)。
//
// Checkpoint = Turn 边界 Snapshot + 校验元数据。
// 恢复:Latest → LoadSnapshot → LoadFrom(seq)(返回 seq 之后的增量)→ Apply。
package state

import (
	"fmt"
	"sync"

	"agent-runtime-go/domain"
)

// Checkpoint 是可跨进程恢复的单元。Round 2 = Snapshot + SchemaVersion。
type Checkpoint struct {
	SchemaVersion int // 当前为 1;不匹配则丢弃,全量 replay
	Snapshot      Snapshot
}

const CheckpointSchemaVersion = 1

// CheckpointStore 在 SnapshotStore 之上加一层 schema 校验。
type CheckpointStore interface {
	Latest(sessionID string) (Checkpoint, bool, error)
	Save(sessionID string, cp Checkpoint) error
}

// MemCheckpointStore 直接保存完整 Checkpoint,确保 SchemaVersion 不会在读回时丢失。
type MemCheckpointStore struct {
	mu          sync.Mutex
	checkpoints map[string]Checkpoint
}

func NewMemCheckpointStore() *MemCheckpointStore {
	return &MemCheckpointStore{checkpoints: map[string]Checkpoint{}}
}

func (s *MemCheckpointStore) Latest(sessionID string) (Checkpoint, bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	cp, ok := s.checkpoints[sessionID]
	if !ok {
		return Checkpoint{}, false, nil
	}
	cp.Snapshot = cloneSnap(cp.Snapshot)
	return cp, true, nil
}

func (s *MemCheckpointStore) Save(sessionID string, cp Checkpoint) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if cp.SchemaVersion == 0 {
		cp.SchemaVersion = CheckpointSchemaVersion
	}
	cp.Snapshot = cloneSnap(cp.Snapshot)
	if existing, ok := s.checkpoints[sessionID]; ok && existing.Snapshot.Seq > cp.Snapshot.Seq {
		return fmt.Errorf("checkpoint save rejected: existing seq %d > new seq %d",
			existing.Snapshot.Seq, cp.Snapshot.Seq)
	}
	s.checkpoints[sessionID] = cp
	return nil
}

// RecoverableState 是恢复所需的最小 State 能力。
type RecoverableState interface {
	State
	LoadSnapshot(sessionID string, view domain.SessionView)
	ResetSession(sessionID string)
}

// Recover 从 Checkpoint + EventStore 增量重建 State。
// 若 Checkpoint schema 不匹配或缺失,先 Reset 再 fromSeq=0 全量 Load。
func Recover(sessionID string, cps CheckpointStore, store EventStore, st RecoverableState) (replayed int, err error) {
	cp, ok, err := cps.Latest(sessionID)
	if err != nil {
		return 0, err
	}
	fromSeq := int64(0)
	if ok && cp.SchemaVersion == CheckpointSchemaVersion {
		if cp.Snapshot.Seq != cp.Snapshot.View.MaxSeq {
			return 0, fmt.Errorf("checkpoint seq/view mismatch: seq=%d maxSeq=%d",
				cp.Snapshot.Seq, cp.Snapshot.View.MaxSeq)
		}
		st.ResetSession(sessionID)
		st.LoadSnapshot(sessionID, cp.Snapshot.View)
		fromSeq = cp.Snapshot.Seq
	} else {
		st.ResetSession(sessionID)
	}
	remaining, err := store.LoadFrom(sessionID, fromSeq)
	if err != nil {
		return 0, err
	}
	if err := st.Apply(remaining); err != nil {
		return 0, fmt.Errorf("recover apply: %w", err)
	}
	return len(remaining), nil
}

// TakeCheckpoint 在 Turn 边界从当前 View 拍照。
func TakeCheckpoint(sessionID string, st State, cps CheckpointStore) error {
	view, err := st.View(sessionID)
	if err != nil {
		return err
	}
	return cps.Save(sessionID, Checkpoint{
		SchemaVersion: CheckpointSchemaVersion,
		Snapshot:      Snapshot{Seq: view.MaxSeq, View: view},
	})
}

File: runtime-rs/src/state/checkpoint.rs
//! Checkpoint 与恢复 —— 与 `runtime-go/state/checkpoint.go` 对齐。见 ch09。

use std::collections::HashMap;

use super::{EventStore, Snapshot, State, StateError};
use crate::domain::SessionView;

pub const CHECKPOINT_SCHEMA_VERSION: i32 = 1;

#[derive(Debug, Clone, Default)]
pub struct Checkpoint {
    pub schema_version: i32,
    pub snapshot: Snapshot,
}

pub trait CheckpointStore {
    fn latest(&self, session_id: &str) -> Result<Option<Checkpoint>, StateError>;
    fn save(&mut self, session_id: &str, cp: Checkpoint) -> Result<(), StateError>;
}

pub struct MemCheckpointStore {
    checkpoints: HashMap<String, Checkpoint>,
}

impl MemCheckpointStore {
    pub fn new() -> Self {
        Self {
            checkpoints: HashMap::new(),
        }
    }
}

impl Default for MemCheckpointStore {
    fn default() -> Self {
        Self::new()
    }
}

impl CheckpointStore for MemCheckpointStore {
    fn latest(&self, session_id: &str) -> Result<Option<Checkpoint>, StateError> {
        Ok(self.checkpoints.get(session_id).cloned())
    }

    fn save(&mut self, session_id: &str, mut cp: Checkpoint) -> Result<(), StateError> {
        if cp.schema_version == 0 {
            cp.schema_version = CHECKPOINT_SCHEMA_VERSION;
        }
        if let Some(existing) = self.checkpoints.get(session_id) {
            if existing.snapshot.seq > cp.snapshot.seq {
                return Err(StateError(format!(
                    "checkpoint save rejected: existing seq {} > new seq {}",
                    existing.snapshot.seq, cp.snapshot.seq
                )));
            }
        }
        self.checkpoints.insert(session_id.into(), cp);
        Ok(())
    }
}

pub trait RecoverableState: State {
    fn load_snapshot(&mut self, session_id: &str, view: SessionView);
    fn reset_session(&mut self, session_id: &str);
}

pub fn take_checkpoint(
    session_id: &str,
    state: &dyn State,
    cps: &mut dyn CheckpointStore,
) -> Result<(), StateError> {
    let view = state.view(session_id)?;
    cps.save(
        session_id,
        Checkpoint {
            schema_version: CHECKPOINT_SCHEMA_VERSION,
            snapshot: Snapshot {
                seq: view.max_seq,
                view,
            },
        },
    )
}

pub fn recover(
    session_id: &str,
    cps: &dyn CheckpointStore,
    store: &dyn EventStore,
    state: &mut dyn RecoverableState,
) -> Result<usize, StateError> {
    let mut from_seq = 0i64;
    if let Some(cp) = cps.latest(session_id)? {
        if cp.schema_version == CHECKPOINT_SCHEMA_VERSION {
            if cp.snapshot.seq != cp.snapshot.view.max_seq {
                return Err(StateError(format!(
                    "checkpoint seq/view mismatch: seq={} max_seq={}",
                    cp.snapshot.seq, cp.snapshot.view.max_seq
                )));
            }
            state.reset_session(session_id);
            state.load_snapshot(session_id, cp.snapshot.view);
            from_seq = cp.snapshot.seq;
        } else {
            state.reset_session(session_id);
        }
    } else {
        state.reset_session(session_id);
    }
    let remaining = store.load_from(session_id, from_seq)?;
    let n = remaining.len();
    state.apply(&remaining)?;
    Ok(n)
}
